### ISSUE TITLE
3.2 commandline -f repaint changes break async prompts on linux systems

### ISSUE THREAD

**Issue by IlanCosman** [fish-shell/fish-shell#7671]
`fish, version 3.1.2-1988-gb8920d493`

The following are defined in the `functions/fish_prompt.fish` file:
```fish
function fish_prompt
    fish --command="sleep .1;set -U foo (pwd)" &
    echo $foo
end

function on_foo --on-variable foo
    commandline -f repaint
end
```

Here's it working as intended on fish 3.1.2 on Arch:

![repaint_3 1 2](https://user-images.githubusercontent.com/6318351/106371576-dd6ade00-631a-11eb-8d52-366aa1c2e488.gif)

Here's fish built from source on Arch. It only updates on an `ENTER` press.

![repaint_built_from_source_arch](https://user-images.githubusercontent.com/6318351/106371578-e360bf00-631a-11eb-8d79-5203f537829c.gif)

Here's `fish, version 3.1.2-1988-gb8920d493` installed from PPA on Ubuntu 20.04. It updates when you press any key.

![repaint_ubuntu_bug](https://user-images.githubusercontent.com/6318351/106371579-e65baf80-631a-11eb-9ef2-942112b80b37.gif)

On macOS, the same version as on Ubuntu works correctly.

**Comment by faho** [comment]
Ah, lovely, it's down to the new SIGIO universal variable notifier strategy.

If I disable that, it works again.

Assigning @ridiculousfish.

**Comment by faho** [comment]
Ah, okay, with the SIGIO notifier you get changes after keypresses, but with the named pipe you don't always get the events either - I just got (un)lucky in my first attempts.

Bisecting...

As a sidenote: "async prompts" have become a lovely little cottage industry, but from our end they're basically unsupportable. It's all hacks upon hacks, and wasn't ever really "meant" to work. So it's unsurprising that it would break.

**Comment by faho** [comment]
Bisects to d5d09c993e1c569ec4d5ba4cb24ff706798382eb - i.e. the fix for #7559.

**Comment by IlanCosman** [comment]
You mean that https://github.com/fish-shell/fish-shell/commit/d5d09c993e1c569ec4d5ba4cb24ff706798382eb is the commit that caused this issue? I'm almost certain that it goes back further than that, but perhaps I'm remembering incorrectly.

I'm really sorry to cause trouble for you Fabian, and I know you counseled against async prompts. But the market has spoken to a certain extent. Perhaps the wise options are either to tell us async users to kick rocks, or to make real concurrence a priority and just get it done. A low level of support may end up costing more energy in the long-term. 🤷‍♂️It's a tough situation.

**Comment by faho** [comment]
>I'm almost certain that it goes back further than that, but perhaps I'm remembering incorrectly.

You are. Check out d5d09c9, it won't repaint until you press a button. Check out d5d09c9^ (the parent commit) and it will.

That's a clean rebase.

**Comment by ridiculousfish** [comment]
Nice catch! The issue is that select() returns with -1 / EINTR instead of a normal return and we weren't polling in that case. Uvar notifiers continue to be a source of pain. Anyways fixed as 2d78c9a0d97d4853064191444ebaa1b2b4425ad0
